### ISSUE TITLE
[REQUEST] Derivative of electron repulsion. 

### ISSUE THREAD

**Issue by alexm-gc** [theochem/gbasis#128]
Your function [electron_repulsion_integral](https://github.com/theochem/gbasis/blob/master/gbasis/integrals/electron_repulsion.py) agrees with PySCF:

```
import pyscf 
import gbasis

mol        = pyscf.gto.M(atom="H 0 0 0; H 0 0 1; ")
pyscf_eri  = mol.intor("int2e")
gbasis_eri = gbasis.electron_repulsion.electron_repulsion_integral(gbasis.wrappers.from_pyscf(mol))

assert np.allclose( pyscf_eri, gbasis_eri ) # this is True
```

I need the derivative of the two electron integral which in PySCF is: 

```
pyscf_eri_deriv = mol.intor("int2e_ip1")
```

I've rewritten your [electron_repulsion_integral](https://github.com/theochem/gbasis/blob/master/gbasis/integrals/electron_repulsion.py) from numpy to [jax.numpy](https://jax.readthedocs.io/en/latest/jax.numpy.html) so I can get the Jacobian in one line of code: 

```
import jax
J_eri = jax.jacfwd( jax_electron_repulsion_integral ) ( atom_positions )
```

**The problem:** `pyscf_eri_deriv` and `J_eri` almost perfectly agree, but there are two entries that are different. 

**Feature Request:** An implementation of the derivative of `electron_repulsion_integral` such that the output is `np.allclose` to `mol.intor("int2e_ip1")` from PySCF.

**Note:** Using `jax.jacfwd` I managed to get jax.numpy implementations of your code to agree with pyscf (which just uses libcint) for 
- gbasis.overlap.overlap_integral
- gbasis.kinetic_energy.kinetic_energy.integral
- gbasis.point_charge.point_charge_integral

**Comment by PaulWAyers** [comment]
I really don't think this is the right way to compute derivatives (certainly not first-derivatives/Jacobians) with respect to nuclear positions for integrals. It should work, but my guess (do you have/know any data on this?) is that its performance is inferior to a direct/conventional implementation. Implementing nuclear gradients of the integrals is on our to-do list, but we haven't gotten around to it yet (long story). 

Trying to figure out how to debug this seems tricky to me. I wouldn't really be sure where to start, presuming the `gbasis` integrals are correct. It basically means that the automatic differentiation differs from the derivative evaluated by finite differences, which seems more like a problem with `jax` than with `gbasis`. (Which is not to say I'm sure about that; it may be that `gbasis` has a bug. Our 2-electron integrals are very much a prototype at this stage, and there is a lot of performance left on the table that I need to get around to extracting sometime. But for this "bug" I wouldn't know where to start to look.) I would be happy to discuss with you further, however. 

Until the term is over, pull-request-reviewing is going *very* slow, so be aware of that.

**Comment by alexm-gc** [comment]
>Trying to figure out how to debug this seems tricky to me.

Apologies for not writing, made `jax.grad( electron_repulsion )` agree with PySCF yesterday. No need to take further action.

> It basically means that the automatic differentiation differs from the derivative evaluated by finite differences, which seems more like a problem with jax than with gbasis

Jax.grad agress with finite differences. The problem was that `jax.grad( gbasis.electron_repulsion )`  disagreed with PySCF/libcint  (this is now resolved, looks to me like a bug in libcint). 

>I really don't think this is the right way to compute derivatives

By "this" are you referring to automatic differentiation? If so, I agree that automatic differentiation is not the conventional approach, but I haven't seen any first-principle arguments against using automatic differentiation. Please do share any first-principle arguments against using automatic differentiation you may have. 

> my guess (do you have/know any data on this?) is that its performance is inferior to a direct/conventional implementation.

That's a great hypothesis, it has the important property that it's falsifiable. 

In general, automatic differentiation `jax.grad( func )` takes 2.5x longer than evaluating `func`; this is also true for our special case where `func=electron_repulsion`. For my application this is sufficiently fast, because jax can compile `jax.grad( electron_repulsion  )` to hardware accelerators which are ~80x faster than CPUs normalized for $. 

If one can write code that computes the derivative of `electron_repulsion` 2.5x faster than evaluating `electron_repulsion` I won't be able to falsify your hypothesis. However, if the derivative takes more than 2.5x longer than evaluating `electron repulsion` the hypothesis "the performance of automatic differentiation is inferior to direct implementation" would be falsified. 

> Implementing nuclear gradients of the integrals is on our to-do list, but we haven't gotten around to it yet (long story).

I got the gradient working yesterday at the cost of one day of work (i.e. it was a very short story). I think this may be the main advantage of automatic differentiation.

**Comment by alexm-gc** [comment]
> I would be happy to discuss with you further, however.

All working, but happy to do a 30 min vchat. 

Unrelated question. Do you have a paper we can cite when using gbasis in publications?

**Comment by PaulWAyers** [comment]
Thanks. We'd certainly welcome a pull request with your new features.

When you say the automatic differentiation is 2.5x slower than the function evaluation, this is 2.5x *per* dimension or 2.5x total. If it were independent of dimension, then a lot of things that we normally do not do in quantum chemistry (e.g., optimizing all the Gaussian exponents in the basis in order to ensure exact satisfaction of the virial theorem) would be (semi)feasible.

The cost of evaluating Gaussian integrals (with differentiation) requires evaluating auxiliary functions, which raise the effective angular momentum of the basis by one. This forces you to go one layer deeper into the recursion; there is then some work associated with piecing these raw ingredients together into the integrals. The cost is not too high, but 2.5x is not dissimilar probably. It would be an interesting comparison to do.

**Comment by PaulWAyers** [comment]
We don't have a paper for `gbasis` yet. 

Would be happy to set up a chat for next week. Just e-mail me. You can find my e-mail on the [QC-Devs web site](https://qcdevs.org/). (Rather not post it willy-nilly everywhere on the internet, obviously.)

**Comment by alexm-gc** [comment]
>  this is 2.5x per dimension or 2.5x total.

It's 2.5x total and **!NOT!** 2.5x per dimension as the case of finite difference. If you're curious this [blog post](https://justindomke.wordpress.com/2009/02/17/automatic-differentiation-the-most-criminally-underused-tool-in-the-potential-machine-learning-toolbox/) clarifies. 

I think this is likely the most hurtful misconception of automatic differentiation; if you understand this misconception and explain it to all your colleagues the entire world will be a much better place. 

> (e.g., optimizing all the Gaussian exponents in the basis in order to ensure exact satisfaction of the virial theorem) would be (semi)feasible.

Yup, computing `d any_function / d gaussian_exponents` is super easy with automatic differentiation, would probably take 30 min to implement in my DFT code base. 

I'll be brutally honest and say I don't think gaussians or plane-waves are suitable for 2022 hardware accelerators. I appreciate people spent decades researching these techniques, but in my experience the best technique often change when one gets 100x more computational power. 

I think as researchers this should be a source of excitement. The field might open up for new completely crazy ideas; a new PhD student somewhere might find a new amazing way of representing electron densities that improves the performance of DFT! 

I'll end by thanking you for this code; it's invaluable to have implementations like this.
